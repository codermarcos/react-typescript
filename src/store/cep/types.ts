export interface Address {
  cep: string;
  logradouro: string;
  complemento: string;
  bairro: string;
  localidade: string;
  uf: string;
  unidade: string;
  ibge: string;
  gia: string;
}

export interface CepState {
  address: Address;
}

export enum CepActionType {
  fetch = 'fetch',
  save = 'save',
}


export interface FetchCepAction {
  type: CepActionType.fetch;
  cep: string;
}

export interface SaveCepAction {
  type: CepActionType.save;
  payload: {
    address: Address;
  };
}

export type CepActionTypes = FetchCepAction | SaveCepAction;
