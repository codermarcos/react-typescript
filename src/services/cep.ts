const baseUrl = 'https://viacep.com.br/ws/';

const cepService = (cep: string) => fetch(`${baseUrl}${cep}/json`).then((response) => response.json());

export default cepService;
