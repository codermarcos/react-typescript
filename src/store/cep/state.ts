import { CepState } from './types';

const state: CepState = {
  address: {
    cep: '',
    logradouro: '',
    complemento: '',
    bairro: '',
    localidade: '',
    uf: '',
    unidade: '',
    ibge: '',
    gia: '',
  },
};

export default state;
