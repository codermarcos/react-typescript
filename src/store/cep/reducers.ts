import { Reducer } from 'react';

import defaultState from 'src/store/cep/state';
import { CepState, CepActionTypes, CepActionType } from 'src/store/cep/types';

const cepReducer: Reducer<CepState, CepActionTypes> = (state = defaultState, action) => {
  switch (action.type) {
    case CepActionType.save:
      return { ...state, ...action.payload };
    default:
      return state;
  }
};

export default cepReducer;
