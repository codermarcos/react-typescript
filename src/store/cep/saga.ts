import { FetchCepAction, CepActionType } from 'src/store/cep/types';
import { call, put, select } from 'redux-saga/effects';
import { push } from 'connected-react-router';
import cepService from 'src/services/cep';
import { RootState } from 'src/store';

function* searchCep(action: FetchCepAction) {
  try {
    const address = yield call(cepService, action.cep);

    yield put({ type: CepActionType.save, payload: { address } });

    const page = yield select((state: RootState) => state.router.location.pathname);

    if (!page || page === '/') yield put(push(action.cep));
  } catch (error) {
    console.log(error);
  }
}

export default searchCep;
