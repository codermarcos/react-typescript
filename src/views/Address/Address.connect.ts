import { connect, ConnectedProps } from 'react-redux';
import { CepActionType } from 'src/store/cep/types';
import { RootState } from 'src/store';

const mapState = (state: RootState) => ({
  address: state.cep.address,
});

const mapDispatch = {
  searchCep: (cep: string) => ({ type: CepActionType.fetch, cep }),
};

const connector = connect(mapState, mapDispatch);

export default connector;

export type PropsFromRedux = ConnectedProps<typeof connector>;
