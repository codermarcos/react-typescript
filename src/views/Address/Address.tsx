import React from 'react';
import { useParams } from 'react-router-dom';

import 'src/views/Address/Address.css';

import { PropsFromRedux } from 'src/views/Address/Address.connect';


function Address(props: PropsFromRedux) {
  let html: JSX.Element = <main>Procurando</main>;
  const { address } = props;

  if (address && address.logradouro) {
    html = (
      <main>
        {address.logradouro}
      </main>
    );
  }

  const { cep } = useParams();

  if (!cep) return html = <main>Cep n√£o encontrado</main>;

  if (cep && cep !== address.cep.replace('-', '')) {
    const { searchCep } = props;
    searchCep(cep);
  }

  return html;
}

export default Address;
