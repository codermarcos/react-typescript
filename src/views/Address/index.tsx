import React from 'react';
import { Helmet } from 'react-helmet';
import { Route, RouteProps } from 'react-router-dom';

import Address from 'src/views/Address/Address';
import connector from 'src/views/Address/Address.connect';

const Connected = connector(Address);

export default (props: RouteProps): JSX.Element => (
  <Route {...props}>
    <Helmet>
      <title>Address | Cep</title>
      <meta name="description" content="Welcome cruiser" />
    </Helmet>
    <Connected />
  </Route>
);
