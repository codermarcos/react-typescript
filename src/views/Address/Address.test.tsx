import React from 'react';
import renderer from 'react-test-renderer';

import Address from 'src/views/Address/Address';
import { PropsFromRedux } from 'src/views/Address/Address.connect';
import { CepActionType } from 'src/store/cep/types';
import { MemoryRouter } from 'react-router-dom';

describe(
  `component: ${Address.name}`,
  () => {
    const props: PropsFromRedux = {
      searchCep(cep: string) {
        return { type: CepActionType.fetch, cep };
      },
      address: {
        cep: '09390-500',
        logradouro: '',
        complemento: '',
        bairro: '',
        localidade: '',
        uf: '',
        unidade: '',
        ibge: '',
        gia: '',
      },
    };

    test(
      'renders correctly',
      () => {
        const component = (
          <MemoryRouter initialEntries={['/09390500']}>
            <Address {...props} />
          </MemoryRouter>
        );

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
      },
    );
  },
);
