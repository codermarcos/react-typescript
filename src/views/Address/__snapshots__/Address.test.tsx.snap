// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: Address renders correctly 1`] = `
<main>
  Cep n√£o encontrado
</main>
`;
